glean-core/python/glean/glean.py

```diff
@@ -315,32 +297,35 @@ def set_upload_enabled(cls, enabled: bool) -> None:
         Args:
             enabled (bool): When True, enable metric collection.
         """
-        if cls.is_initialized():
+        # Changing upload enabled always happens asynchronous.
+        # That way it follows what a user expect when calling it inbetween other calls:
+        # It executes in the right order.
+        #
+        # Because the dispatch queue is halted until Glean is fully initialized
+        # we can safely enqueue here and it will execute after initialization.
+        @Dispatcher.launch
+        def set_upload_enabled():
             original_enabled = cls.get_upload_enabled()
+            _ffi.lib.glean_set_upload_enabled(enabled)
 
-            @Dispatcher.launch
-            def set_upload_enabled():
-                _ffi.lib.glean_set_upload_enabled(enabled)
-
-                if original_enabled is False and cls.get_upload_enabled() is True:
-                    cls._initialize_core_metrics()
-
-                if original_enabled is True and cls.get_upload_enabled() is False:
-                    # If uploading is disabled, we need to send the deletion-request ping
-                    PingUploadWorker.process()
+            if original_enabled is False and cls.get_upload_enabled() is True:
+                cls._initialize_core_metrics()
 
-        else:
-            cls._upload_enabled = enabled
+            if original_enabled is True and cls.get_upload_enabled() is False:
+                # If uploading is disabled, we need to send the deletion-request ping
+                PingUploadWorker.process()
 
     @classmethod
     def get_upload_enabled(cls) -> bool:
```

Comment:
Should we simply remove this/change this to `_get_upload_enabled` so that the intent is clearer? I don't think anybody is using this.

Reply:
👍